### PR TITLE
Update the snapshots script to use combined examples if they're found

### PR DESCRIPTION
**Pull request by jmuzina**
## Done

First segment of split-up of #5117 as requested by @pastelcyborg .

This PR only includes the changes to `snapshots.js` to make it snapshot `combined.html` templates where found, and take snapshots in every color theme on every example. Note that until combined examples themselves are merged, this script would mean a dramatic increase in snapshots per build as we are taking snapshots in each theme now.

It also cleans up `snapshots.js` slightly to make sure it is well-documented, and adds a set of Jest tests to verify correctness of the Snapshots script.

## QA

- Verify PR test passes (it will run a Jest test against `snapshots.test.js` verifying correctness of the Percy script)

### Check if PR is ready for release

If this PR contains Vanilla SCSS code changes, it should contain the following changes to make sure it's ready for the release:

- [x] PR should have one of the following labels to automatically categorise it in release notes:
  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.
- [ ] Vanilla version in `package.json` should be updated relative to the [most recent release](https://github.com/canonical/vanilla-framework/releases/latest), following semver convention:
  - if CSS class names are not changed it can be bugfix relesase (x.x.**X**)
  - if CSS class names are changed/added/removed it should be minor version (x.**X**.0)
  - see the [wiki for more details](https://github.com/canonical/vanilla-framework/wiki/Release-process#pre-release-tasks)
- [ ] Any changes to component class names (new patterns, variants, removed or added features) should be listed on the [what's new page](https://github.com/canonical/vanilla-framework/blob/main/releases.yml).

### CODE REVIEW

File: tests/snapshots.test.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
+const {URL} = require('url');
+const snapshotsTest = require('../snapshots');
+
+const PORT = process.env.PORT || 8101;
+
+test('Number of snapshots', async () => {
+  const snapshots = await snapshotsTest();
+  const totalSnapshots = snapshots.reduce((acc, snapshot) => acc + snapshot.widths.length, 0);
+  expect(totalSnapshots).toBe(2906);
+});
```
Comment (jmuzina) [inline tests/snapshots.test.js:10]: This test would need to be updated every time the total number of snapshots taken by Percy changes, so we may find ourselves updating it frequently. I'll probably remove this but just noting it for the rest of the team.

File: snapshots.js [hunk 2]
```diff
@@ -2,45 +2,134 @@ const fs = require('fs');
...
   });
 }
 
+/**
+ * Given an example path, returns the color themes to snapshot.
+ * @param filepath {String} Path (relative to `EXAMPLES_RELATIVE_DIR`) of the example file.
+ * @returns {String[]} Array of color theme names to snapshot.
+ *  These must match color theme query parameter names accepted by example-tools.js.
+ */
+function getThemesForExample(filepath) {
```
Comment (bartaz) [inline snapshots.js:69]: What's the point of that function? The filepath param is not used at all.

Is this a legacy of some config used before, or preparation for adding something later?

File: snapshots.js [hunk 3]
```diff
@@ -2,45 +2,134 @@ const fs = require('fs');
...
+ * @param filepath {String} Path (relative to `EXAMPLES_RELATIVE_DIR`) of the example file.
+ * @returns {Promise<Number[]>} Array of widths to snapshot, from smallest to largest.
+ */
+async function getWidthsForExample(filepath) {
+  let widths = new Set([
+    375, // Mobile
+    1280, // Desktop
+  ]);
+
+  if (filepath.endsWith('combined/')) {
```
Comment (bartaz) [inline snapshots.js:84]: I don't exactly understand what is the purpose of this code. What changes if the combined path is provided?
Isn't "combined" just an HTML file? In here it seems to be treated as directory.

File: tests/snapshots.test.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
+const {URL} = require('url');
+const snapshotsTest = require('../snapshots');
+
+const PORT = process.env.PORT || 8101;
+
+test('Snapshot widths', async () => {
```
Comment (bartaz) [inline tests/snapshots.test.js:6]: I prefer a more descriptive "behavioural" test names that explain what the tested code is expected to do.
This helps in understanding the test code, but also makes sure tests are written from perspective of expected results.

In here I would say something like:
`test("returns correct widths for snapshots, including additional one for responsive examples", … )`
or something like that.

File: tests/snapshots.test.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      }
+    }
+    expectedWidths = Array.from(expectedWidths).sort((a, b) => a - b);
+
+    return JSON.stringify(snapshot.widths) !== JSON.stringify(expectedWidths);
+  });
+  expect(failedUrls).toHaveLength(0);
+});
+
+test('Unique snapshot names', async () => {
```
Comment (bartaz) [inline tests/snapshots.test.js:27]:
```suggestion
test('returns snapshots with unique names', async () => {
```

**Review by bartaz — COMMENTED** [review]
I added few questions/comments in the code. Let's clarify those and it should be good to go.

File: snapshots.js (hunk 3 above)
Comment (jmuzina) [reply snapshots.js]: Changed param names & added documentation to make this more clear, let me know if that is sufficient @bartaz

**Review by bartaz — APPROVED** [review]
Thanks, let's start combining them!

**Pull request merged**